src/layout.rs

```diff
@@ -116,6 +116,22 @@ impl Constraint {
             Constraint::Min(m) => length.max(m),
         }
     }
+
+    pub(crate) fn to_cassowary(self, lhs: Variable, rhs: Variable) -> Vec<cassowary::Constraint> {
+        match self {
+            Constraint::Percentage(p) => vec![lhs | EQ(STRONG) | (rhs * f64::from(p) / 100.00)],
+            Constraint::Ratio(n, d) => vec![lhs | EQ(STRONG) | (rhs * f64::from(n) / f64::from(d))],
+            Constraint::Length(l) => vec![lhs | EQ(STRONG) | f64::from(l)],
+            Constraint::Max(m) => vec![
+                lhs | LE(STRONG) | f64::from(m),
+                lhs | EQ(WEAK) | f64::from(m),
```

Comment:
> `lhs | LE(STRONG) | f64::from(m),`

By using `STRONG` it is not guaranteed that this constraint will be met.

> `lhs | EQ(WEAK) | f64::from(m),`

I think by adding this you incentivize the algorithm to be closer to the value in `Max`, i.e. `m`.

I'm not sure if both of these are what users would expect?

I personally would prefer the following:

```
lhs | LE(REQUIRED) | f64::from(m),
```

i.e. just one constraint. 

Maybe we should add another `Constraint::ApproxMax`?

Reply:
This PR is intended to be a pure refactor without changing any behavior. This is the current way that the min and max constraints work. I don't think it is necessarily right, and agree we should discuss how this should work.
https://github.com/ratatui-org/ratatui/blob/ea48af1c9abac7012e3bf79e78c6179f889a6321/src/layout.rs#L419-L428

The reason that this isn't required is to avoid panics when the user supplied values can't be satisfied.
Contrast the required constraints that can always be satisfied (area size, all chunks within the area, all chunks laid out left to right / top to bottom in order, and the left / top touching the start).

Another way of thinking about this is why should `Max` be more special than `Length` or `Min` if we have asked the area to expand to fill, and if it's not, then those should be required.

Reply:
My bad, I misread / misremembered what the original was doing, and I didn't realize until your comment this PR didn't change any behavior.

> we should discuss how this should work.

I agree. I'm working on something that might help with this. I'll discuss in a separate thread.